### PR TITLE
Fix the multicore performance of mapping from an instruction pointer to a MethodDesc

### LINKED ISSUE #8393: ExecutionManager::GetRangeSection performs a linear search over list of loaded assemblies
PerfView indicates that significant time can be spent in this method when  we create a Delegate:
The MusicStore application has over 130 managed assemblies loaded.
Eliminating this search can improve steady state transaction time by ~3%

|+ coreclr!COMDelegate::DelegateConstruct
We have to perform two IP lookups in DelegateConstruct

```
       FCIMPL3(void, COMDelegate::DelegateConstruct, Object* refThisUNSAFE, Object* targetUNSAFE, PCODE method)
(line 1914)
219.0 |    MethodDesc * pCreatorMethod = ExecutionManager::GetCodeMethodDesc((PCODE)pRetAddr);
(line 1933)
159.0 |    MethodDesc *pMethOrig = Entry2MethodDesc(method, pRealMT);
```
Both of these calls eventually resolve to: ExecutionManager::GetRangeSection 
```
RangeSection* ExecutionManager::GetRangeSection(TADDR addr)
. . . 
          while (pCurr != NULL)
           {
               // See if addr is in [pCurr->LowAddress .. pCurr->HighAddress)
 12.0 |        if (pCurr->LowAddress <= addr)
               {
                   // Since we are sorted, once pCurr->HighAddress is less than addr
                   // then all subsequence ones will also be lower, so we are done.
  3.0 |            if (addr >= pCurr->HighAddress)
                   {
                       // we'll return NULL and put pLast into pLastUsed
                       pCurr = NULL;
                   }
                   else
                   {
                       // addr must be in [pCurr->LowAddress .. pCurr->HighAddress)
                       _ASSERTE((pCurr->LowAddress <= addr) && (addr < pCurr->HighAddress));
       
                       // Found the matching RangeSection
                       // we'll return pCurr and put it into pLastUsed
                       pLast = pCurr;
                   }
       
                   break;
               }
320.0 |        pLast = pCurr;
 57.0 |        pCurr = pCurr->pnext;
           }
```

### PR DESCRIPTION
**Pull request by davidwrighton**
Mapping from an instruction pointer to a MethodDesc is a critical component of the runtime used in many places, notably diagnostics, slow path delegate creation, and stackwalking.

This change provides a dramatic improvement in the performance of that logic through several techniques.

1. The mapping from IP to range of interesting memory is done via a tree structure resembling that of a page table. The result of this is that in addition to reducing the locking logic, the cost of looking up in the presence of many different loaded assemblies will be reduced to a nearly constant time.
2. That tree structure is configured so that when accessing memory which will never be freed in the lifetime of the process, the memory can be accessed without taking any locks whatsoever.
3. The code map is enhanced to include not only the code generated by the JIT/R2R code, but also to include the fixup and stub precode stubs.
4. In addition, performance improvement was done to improve the performance of slow path delegate creation in particular, by reordering which checks are done, and by writing a simplified signature parse routine for computing the number of arguments to a function.

Performance of this was tested in both the EH stackwalking scenario as well as the delegate slow path creation scenario, but the delegate logic is what will be most visible when this is checked in. (See PR #79471 for details of the additional changes necessary to take advantage of this work when doing EH) (#8393 describes the potential product impact of just improving the delegate slow path creation code)

For the delegate creation scenario the perf impact was measured to be quite substantial. These numbers are in ms to create a constant number of delegates on each thread used. Smaller is better. The test was run on a 6 core machine.

| Threads | Without fix | With this PR
| ------------- | ------------- | ----- |
| 1  | 840 |  313|
| 2 | 1977 | 316 |
| 3 | 3290 | 325 |
| 4| 4259 | 344 |
| 5 | 5140 | 351 |
| 6 | 5872 | 374|
| 7 | 6463 | 442|
| 8 | 7046 | 499|
| 9 | 7648 | 547|
| 10 | 8241 | 627|
| 11 | 8851 | 749|
| 12 | 9595 | 773|

Fixes #8393
Replaces the controversial logic in PR #68607

### CODE REVIEW

**Comment by EgorBo** [comment]
Nice! Does it make `ExecutionEngine::IsReadyToRun(PCODE)` cheap?

**Comment by davidwrighton** [comment]
@EgorBo, yes, I do expect that this will make IsReadyToRun very cheap. (Probably 5-6 memory loads from separate pages plus some math, so repeated calls should be extremely cheap, and random calls will be somewhat expensive, but not bad.) The implementation as written has a risk of entering a lock under some circumstances, but after sleeping on it, I think we could get rid of that for the IsReadyToRun logic, and be assured that IsReadyToRun is cheap. What do you think you can optimize?

**Comment by EgorBo** [comment]
> What do you think you can optimize?

I'd like to sort of de-prioritize R2R'd code for tier1 promotion in favor of non R2R'd one.

**Comment by davidwrighton** [comment]
@EgorBo, fyi, `ExecutionManager::IsReadyToRun` is slower than `ExecutionManager::FindReadyToRunModule`. Also, I've found there may be a slight bug in the current implementation of the locking strategy for those 2 methods, so I'm addressing that as well.

File: src/coreclr/vm/codeman.h [hunk 1]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+    // Helper structure which forces all access to the various pointers to be handled via volatile/interlocked operations
+    // The copy/move constructors are all deleted to forbid accidental reads into temporaries, etc.
+    class RangeSectionFragmentPointer
+    {
+    private:
+        TADDR _ptr;
+
+        TADDR FragmentToPtr(RangeSectionFragment* fragment)
+        {
+            TADDR ptr = (TADDR)fragment;
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:787]:
```suggestion
            TADDR ptr = dac_cast<TADDR>(fragment);
```
or ifdef out the method for DACCESS_COMPILE

File: src/coreclr/vm/codeman.h [hunk 2]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+    typedef DPTR(RangeSectionFragmentPointer) PTR_RangeSectionFragmentPointer;
+
+    // Helper structure which forces all access to the various pointers to be handled via volatile/interlocked operations
+    // The copy/move constructors are all deleted to forbid accidental reads into temporaries, etc.
+    class RangeSectionFragmentPointer
+    {
+    private:
+        TADDR _ptr;
+
+        TADDR FragmentToPtr(RangeSectionFragment* fragment)
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:785]:
```suggestion
        static TADDR FragmentToPtr(RangeSectionFragment* fragment)
```

File: src/coreclr/vm/codeman.h [hunk 3]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+
+    // Helper structure which forces all access to the various pointers to be handled via volatile/interlocked operations
+    // The copy/move constructors are all deleted to forbid accidental reads into temporaries, etc.
+    template <class TPtr>
+    class RangeSectionLevelPointer
+    {
+    private:
+        TADDR _ptr;
+
+        TADDR LevelToPtr(TPtr level, bool collectible)
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:858]:
```suggestion
        static TADDR LevelToPtr(TPtr level, bool collectible)
```

File: src/coreclr/vm/codeman.h [hunk 4]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+        {
+            TADDR ptr = (TADDR)level;
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:860]:
```suggestion
            TADDR ptr = dac_cast<TADDR>(level);
```

File: src/coreclr/vm/codeman.h [hunk 5]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+
+        void Uninstall()
+        {
+            ::VolatileStore(&_ptr, (TADDR)0);
+        }
+#endif // DACCESS_COMPILE
+    };
+
+    // Unlike a RangeSection, a RangeSectionFragment cannot span multiple elements of the last level of the RangeSectionMap
+    // Always allocated via memset/free
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:968]:
```suggestion
    // Always allocated via malloc/free
```

File: src/coreclr/vm/codeman.h [hunk 6]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+        bool isPrimaryRangeSectionFragment; // RangeSectionFragment are allocated in arrays, but we only need to free the first allocated one. It will be marked with this flag.
+        bool isCollectibleRangeSectionFragment; // RangeSectionFragments
     };
 
-    DWORD               flags;
+#ifdef TARGET_64BIT
+    static const uintptr_t entriesPerMapLevel = 256;
+#else
+    static const uintptr_t entriesPerMapLevel = 256;
+#endif
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:984]:
```suggestion
    static const uintptr_t entriesPerMapLevel = 256;
```

File: src/coreclr/vm/codeman.h [hunk 7]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+
+    void* AllocateLevel()
+    {
+        size_t size = entriesPerMapLevel * sizeof(void*);
+        void *buf = malloc(size);
+        memset(buf, 0, size);
+        return buf;
+    }
+
+    uintptr_t EffectiveBitsForLevel(TADDR address, uintptr_t level)
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:1026]:
```suggestion
    static uintptr_t EffectiveBitsForLevel(TADDR address, uintptr_t level)
```

File: src/coreclr/vm/codeman.h [hunk 8]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+        if (_RangeSectionL1 == NULL)
+            return NULL;
+
+        return ((*_RangeSectionL1)[EffectiveBitsForLevel(address, 1)]).VolatileLoadWithoutBarrier(pLockState);
+    }
+
+    uintptr_t RangeSectionFragmentCount(PTR_RangeSection pRangeSection)
+    {
+        uintptr_t rangeSize = pRangeSection->_range.RangeSize();
+        if (rangeSize == 0)
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:1163]: How can we end up with RangeSection here?

File: src/coreclr/vm/codeman.h [hunk 9]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+        RangeSectionLockState lockState = RangeSectionLockState::ReaderLocked;
+        EnumMemoryRangeSectionMapLevel(flags, GetTopLevel(), &lockState);
+    }
+#endif// DACCESS_COMPILE
+
+};
+
+struct RangeSectionMapData
+{
+    BYTE Data[sizeof(RangeSectionMap)];
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:1532]:
```suggestion
    TADDR Data[sizeof(RangeSectionMap) / sizeof(TADDR)];
```
We need to make sure that the compiler places the static at sufficiently aligned address.

File: src/coreclr/vm/codeman.h [hunk 10]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+            return 0;
+
+        // Account for the range not starting at the beginning of a last level fragment
+        rangeSize += pRangeSection->_range.RangeStart() & (this->bytesAtLastLevel - 1);
+        
+        uintptr_t fragmentCount = ((rangeSize - 1) / bytesAtLastLevel) + 1;
+        return fragmentCount;
+    }
+
+    TADDR IncrementAddressByMaxSizeOfFragment(TADDR input)
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:1173]:
```suggestion
    static TADDR IncrementAddressByMaxSizeOfFragment(TADDR input)
```

File: src/coreclr/vm/codeman.h [hunk 11]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+    RangeSectionMap() : _pCleanupList(EndOfCleanupListMarker())
+    {
+        memset(&_topLevelData, 0, sizeof(_topLevelData));
+    }
+
+    bool Init()
+    {
+        return true;
+    }
+
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:1288]:
```suggestion
```
Not needed?

File: src/coreclr/vm/codeman.h [hunk 12]
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
...
+        return &((*levelToGetPointerIn)[index]);
+    }
+    // Returns pointer to address in last level map that actually points at RangeSection space.
+    RangeSectionFragmentPointer* EnsureMapsForAddress(TADDR address, bool collectible)
+    {
+        uintptr_t level = mapLevels + 1;
+        uintptr_t topLevelIndex = EffectiveBitsForLevel(address, --level);
+        auto nextLevelAddress = &(GetTopLevel()[topLevelIndex]);
+#ifdef TARGET_64BIT
+        auto _RangeSectionL4 = nextLevelAddress;
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:1075]: Nit: Do not use `_` for locals

**Review by jkotas — APPROVED** [review]
LGTM modulo feedback. Nice work!

**Pull request merged**